lib/read-stream.js

```diff
@@ -246,20 +246,31 @@ ReadStream.prototype = {
       return void this.once( 'open', () => { this._read() })
     }
 
-    if( this.ranges.length === 0 ) {
-      this._verifyRange()
-      return void this.push( null )
-    }
-
-    if( this.currentRange == null || this.currentRange.offset === this.currentRange.length ) {
-      this._verifyRange()
-      this.currentRange = this.ranges.shift()
-      this.rangesRead++
-      debug( 'read:range %O', this.currentRange )
+    if (this.currentRange == null) {
+      if (this.ranges.length > 0) {
+        this.currentRange = this.ranges.shift()
+        this.rangesRead++
+        debug( 'read:range %O', this.currentRange )
+        this._readBlock()
+      } else {
+        this.push(null)
+      }
+    } else {
+      if (this.currentRange.offset === this.currentRange.length) {
+        this._verifyRange()
+        if (this.ranges.length > 0) {
+          this.currentRange = this.ranges.shift()
+          this.rangesRead++
+          debug( 'read:range %O', this.currentRange )
+          this._readBlock()
+        } else {
+          this.push(null)
+        }
```

Comment:
Looks like you've got two identical code-blocks:
```js
if (this.ranges.length > 0) {
    this.currentRange = this.ranges.shift()
    this.rangesRead++
    debug( 'read:range %O', this.currentRange )
    this._readBlock()
} else {
    this.push(null)
}
```

Is it worth refactoring into a private helper method?

Reply:
I'd rather not. This `_read` method is already split between `_read` and `_readBlock`, this won't help understanding what it does.

Reply:
Looks like this could also be written as the following, to avoid code duplication;

```js
var hasCurrentRange = this.currentRange != null
var isRangeEnd = hasCurrentRange &&
  ( this.currentRange.offset === this.currentRange.length )

if( !hasCurrentRange || isRangeEnd ) {
  if ( isRangeEnd ) {
    this._verifyRange()
  }
  if( this.ranges.length > 0 ) {
    this.currentRange = this.ranges.shift()
    this.rangesRead++
    debug( 'read:range %O', this.currentRange )
    this._readBlock()
  } else {
    this.push(null)
  }
} else {
  this._readBlock()
}
```

Reply:
@jhermsmeier 
This is a bit harder to read and calls `_verifyRange` even if there is no current range.
If we want to avoid code duplication ,I'd go with @lurch 's suggestion:
```javascript
  /**
   * Advance to next the Range if there is one then read a block;
   * else end the stream;
   * @see https://nodejs.org/api/stream.html#stream_implementing_a_readable_stream
   * @private
   */
  _nextRangeAndReadBlock() {
    if (this.ranges.length > 0) {
      this.currentRange = this.ranges.shift()
      this.rangesRead++
      debug( 'read:range %O', this.currentRange )
      this._readBlock()
    } else {
      this.push(null)
    }
  },

  /**
   * Initiate a new read, advancing the range if necessary,
   * and verifying checksums, if enabled
   * @see https://nodejs.org/api/stream.html#stream_implementing_a_readable_stream
   * @private
   */
  _read() {

    if( this.fd == null ) {
      return void this.once( 'open', () => { this._read() })
    }

    if (this.currentRange == null) {
      this._nextRangeAndReadBlock()
    } else {
      if (this.currentRange.offset === this.currentRange.length) {
        this._verifyRange()
        this._nextRangeAndReadBlock()
      } else {
        this._readBlock()
      }
    }

  },

```

Reply:
> and calls _verifyRange even if there is no current range.

How? `isRangeEnd` is `false` if there is no current range; `hasCurrentRange && ...`

Reply:
You're right, I misread. I still prefer the current version though (the one in the repo now, not the one you reviewed).

Reply:
I think your second version is preferable; it's a lot better to read and understand – only nit would be the name of `_nextRangeAndReadBlock()`, maybe change that to `_readNextRange()`?